packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+    if (this.maintainPeersIntervalId) {
+      this.log.info("Stopping maintain peers interval");
+      clearInterval(this.maintainPeersIntervalId);
+      this.maintainPeersIntervalId = undefined;
+      this.log.info("Maintain peers interval stopped");
+    }
+  }
+
+  /**
+   * Checks if there are peers to send a message to.
+   * If there are no peers, tries to find new peers.
+   * If no peers are found, returns false.
+   * If peers are found, returns true.
+   */
+  protected hasPeers = async (): Promise<boolean> => {
+    let success = await this.maintainPeers();
+    let attempts = 0;
+    while (!success || this.peers.length === 0) {
+      attempts++;
+      success = await this.maintainPeers();
+      if (attempts > 3) {
+        if (this.peers.length === 0) {
+          this.log.error("Failed to find peers to send message to");
+          return false;
+        } else {
+          this.log.warn(
+            `Found only ${this.peers.length} peers, expected ${this.numPeersToUse}`
+          );
+          return true;
+        }
+      }
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    }
+    return true;
+  };
+
+  /**
+   * Starts an interval to maintain the peers list to `numPeersToUse`.
+   * @param interval The interval in milliseconds to maintain the peers.
+   */
+  private async startMaintainPeersInterval(interval: number): Promise<void> {
+    this.log.info("Starting maintain peers interval");
+    try {
+      await this.maintainPeers();
+      this.maintainPeersIntervalId = setInterval(() => {
+        this.maintainPeers().catch((error) => {
+          this.log.error("Error during maintain peers interval:", error);
+        });
+      }, interval);
+      this.log.info(
+        `Maintain peers interval started with interval ${interval}ms`
+      );
+    } catch (error) {
+      this.log.error("Error starting maintain peers interval:", error);
+      throw error;
+    }
+  }
+
+  /**
+   * Maintains the peers list to `numPeersToUse`.
+   */
+  private async maintainPeers(): Promise<boolean> {
+    if (this.maintainPeersLock) {
+      this.log.info("Maintain peers already in progress, skipping");
+      return false;
+    }
+
+    this.maintainPeersLock = true;
+    this.log.info(`Maintaining peers, current count: ${this.peers.length}`);
+    try {
+      const numPeersToAdd = this.numPeersToUse - this.peers.length;
+      if (numPeersToAdd > 0) {
+        await this.findAndAddPeers(numPeersToAdd);
+      }
+      this.log.info(
+        `Peer maintenance completed, current count: ${this.peers.length}`
+      );
+    } catch (error) {
+      this.log.error("Error maintaining peers:", error);
+      throw error;
+    } finally {
+      this.maintainPeersLock = false;
+    }
+    return true;
+  }
+
+  /**
+   * Finds and adds new peers to the peers list.
+   * @param numPeers The number of peers to find and add.
+   */
+  private async findAndAddPeers(numPeers: number = 1): Promise<void> {
```

Comment:
nit: let's have it always provided, unless it is the case that `findAndAddPeers` often called with `1` ofc

Reply:
```suggestion
  private async findAndAddPeers(numPeers: number): Promise<void> {
```

let's have `numPeers` mandatory and not optional